Sources/FoundationEssentials/Subprocess/Subprocess+Configuration.swift

```diff
@@ -0,0 +1,615 @@
+
+            if let outputError = outputError {
+                throw outputError
+            }
+
+            if let errorError = errorError {
+                throw errorError
+            }
+
+            if let killError = killError {
+                throw killError
+            }
+        }
+
+        /// Close each input individually, and throw the first error if there's multiple errors thrown
+        @Sendable
+        internal func cleanupAll(
+            input: ExecutionInput,
+            output: ExecutionOutput,
+            error: ExecutionOutput
+        ) throws {
+            var inputError: Error?
+            var outputError: Error?
+            var errorError: Error?
+
+            do {
+                try input.closeAll()
+            } catch {
+                inputError = error
+            }
+
+            do {
+                try output.closeAll()
+            } catch {
+                outputError = error
+            }
+
+            do {
+                try error.closeAll()
+            } catch {
+                errorError = error
+            }
+
+            if let inputError = inputError {
+                throw inputError
+            }
+            if let outputError = outputError {
+                throw outputError
+            }
+            if let errorError = errorError {
+                throw errorError
+            }
+        }
+
+        internal func run<R>(
+            output: RedirectedOutputMethod,
+            error: RedirectedOutputMethod,
+            _ body: @Sendable @escaping (Subprocess, StandardInputWriter) async throws -> R
+        ) async throws -> Result<R> {
+            let (readFd, writeFd) = try FileDescriptor.pipe()
+            let executionInput: ExecutionInput = .customWrite(readFd, writeFd)
+            let executionOutput: ExecutionOutput = try output.createExecutionOutput()
+            let executionError: ExecutionOutput = try error.createExecutionOutput()
+            let process: Subprocess = try self.spawn(
+                withInput: executionInput,
+                output: executionOutput,
+                error: executionError)
+            // After spawn, cleanup child side fds
+            try self.cleanup(
+                process: process,
+                childSide: true,
+                parentSide: false,
+                attemptToTerminateSubProcess: false
+            )
+            return try await withTaskCancellationHandler {
+                return try await withThrowingTaskGroup(of: RunState<R>.self) { group in
+                    group.addTask {
+                        let status = monitorProcessTermination(
+                            forProcessWithIdentifier: process.processIdentifier)
+                        return .monitorChildProcess(status)
+                    }
+                    group.addTask {
+                        do {
+                            let result = try await body(process, .init(fileDescriptor: writeFd))
```

Comment:
> If we cancel their body closure 
We are not currently canceling the body closure when the subprocess finishes. They run on separate tasks and `run` waits for both to finish.

Reply:
You are right this is very subtle since the monitor task is not throwing. My other point still stands that we should avoid running the users `body` closure in a child task here since it forces the closure to become `@Sendable` and `@escaping` which makes it hard to use in some context e.g. inside an actor.

We should run the closure in the calling task and the monitoring of the child process can happen in a child task but in the end it must surface through the `read/write` operations of the `Subprocess`. From what I can tell this is already happening and you can just move the call to `body` to the body of the `withTaskGroup` here.